bundles/org.openhab.binding.wled/src/main/java/org/openhab/binding/wled/internal/WLedHandler.java

```diff
@@ -0,0 +1,439 @@
+        }
+        return HSBType.fromRGB(r, g, b);
+    }
+
+    private void parseColours(String message) {
+        primaryColor = parseToHSBType(message, "<cl>");
+        updateState(CHANNEL_PRIMARY_COLOR, primaryColor);
+        secondaryColor = parseToHSBType(message, "<cs>");
+        updateState(CHANNEL_SECONDARY_COLOR, secondaryColor);
+        try {
+            primaryWhite = new BigDecimal(WLedHelper.getValue(message, "<wv>", "<"));
+            if (primaryWhite.intValue() > -1) {
+                hasWhite = true;
+                updateState(CHANNEL_PRIMARY_WHITE,
+                        new PercentType(primaryWhite.divide(new BigDecimal(2.55), RoundingMode.HALF_UP)));
+                secondaryWhite = new BigDecimal(WLedHelper.getValue(message, "<ws>", "<"));
+                updateState(CHANNEL_SECONDARY_WHITE,
+                        new PercentType(secondaryWhite.divide(new BigDecimal(2.55), RoundingMode.HALF_UP)));
+            }
+        } catch (NumberFormatException e) {
+            logger.warn("NumberFormatException when parsing the WLED colour and white fields.");
+        }
+    }
+
+    /**
+     *
+     * This function should prevent the need to keep updating the binding as more FX and Palettes are added to the
+     * firmware.
+     */
+    private void scrapeChannelOptions(String message) {
+        List<StateOption> fxOptions = new ArrayList<>();
+        List<StateOption> palleteOptions = new ArrayList<>();
+        int counter = 0;
+        for (String value : WLedHelper.getValue(message, "\"effects\":[", "]").replace("\"", "").split(",")) {
+            fxOptions.add(new StateOption(Integer.toString(counter++), value));
+        }
+        stateDescriptionProvider.setStateOptions(new ChannelUID(getThing().getUID(), CHANNEL_FX), fxOptions);
+        counter = 0;
+        for (String value : (WLedHelper.getValue(message, "\"palettes\":[", "]").replace("\"", "")).split(",")) {
+            palleteOptions.add(new StateOption("" + counter++, value));
+        }
+        stateDescriptionProvider.setStateOptions(new ChannelUID(getThing().getUID(), CHANNEL_PALETTES), palleteOptions);
+    }
+
+    private void processState(String message) {
+        logger.trace("WLED states are:{}", message);
+        if (thing.getStatus() != ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.ONLINE);
+            sendGetRequest("/json"); // fetch FX and Pallete names
+        }
+        if (message.contains("\"effects\":[")) {// JSON API reply
+            scrapeChannelOptions(message);
+            return;
+        }
+        if (message.contains("<ac>0</ac>")) {
+            updateState(CHANNEL_MASTER_CONTROLS, OnOffType.OFF);
+        } else {
+            masterBrightness = new BigDecimal(WLedHelper.getValue(message, "<ac>", "<"));
+            updateState(CHANNEL_MASTER_CONTROLS,
+                    new PercentType(masterBrightness.divide(new BigDecimal(2.55), RoundingMode.HALF_UP)));
```

Comment:
Should probably make `new BigDecimal(2.55)` a constant since you use it in quite a few places.

Reply:
DONE